File: tag.py
# -*- coding: utf-8 -*-
import sys
import time
import csv
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.chrome.options import Options

def search_tag_total(rut):
    # Configurar opciones de Chrome
    chrome_options = Options()
    chrome_options.add_argument("--headless")  # Ejecutar Chrome en modo headless
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")

    # Inicializar el navegador con las opciones configuradas
    driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=chrome_options)

    try:
        # Cargar la primera página
        driver.get("https://unired.tagtotal.cl/")

        # Esperar a que la página cargue completamente
        time.sleep(5)

        # Ingresar el RUT
        rut_input = driver.find_element(By.ID, "rut")
        rut_input.send_keys(rut)

        # Hacer clic en el botón "Buscar"
        search_button = driver.find_element(By.ID, "rut-btn")
        search_button.click()

        # Esperar a que la página cargue los resultados
        time.sleep(20)

        # Obtener las deudas
        deudas = driver.find_elements(By.CSS_SELECTOR, ".row-tabla-deuda")

        # Guardar información en CSV
        with open('tag.csv', 'w', newline='', encoding='utf-8') as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(['Autopista', 'RUT Cliente', 'Tipo Deuda', 'Monto'])
            for deuda in deudas:
                autopista = deuda.find_element(By.CSS_SELECTOR, ".nombre-s").text.strip()
                info = deuda.find_elements(By.CSS_SELECTOR, "p")
                for i in range(0, len(info), 2):
                    rut_cliente = info[i].text.strip()
                    tipo_deuda = info[i+1].text.strip()
                    monto = deuda.find_element(By.CSS_SELECTOR, ".switch.medium + span").text.strip()

                    writer.writerow([autopista, rut_cliente, tipo_deuda, monto])

        # Obtener el valor total a pagar
        total_pagar = driver.find_element(By.CSS_SELECTOR, ".precio-desktop").text
        print("\nTotal a pagar:", total_pagar)

    except Exception as e:
        print("Ocurrió un error:", e)

    finally:
        # Cerrar el navegador al finalizar
        driver.quit()

# Obtener el RUT como argumento de línea de comandos
if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Uso: python tag.py <RUT>")
        sys.exit(1)

    rut = sys.argv[1]
    search_tag_total(rut)

File: tag.js
const { chromium } = require('playwright');
const fs = require('fs');

async function searchTagTotal(rut) {
    const browser = await chromium.launch({ headless: false });
    const page = await browser.newPage();

    try {
        // Navegar a la página principal
        await page.goto('https://unired.tagtotal.cl/');

        // Ingresar el RUT
        await page.fill('input#rut', rut);

        // Hacer clic en el botón "Buscar"
        await Promise.all([
            page.waitForNavigation(),  // Esperar la navegación
            page.click('button#rut-btn')
        ]);

        // Esperar a que la página cargue completamente
        await page.waitForTimeout(15000);  // Aumentar el tiempo de espera si es necesario

        // Obtener las deudas
        const deudas = await page.$$('.row-tabla-deuda');

        // Guardar información en CSV
        const csvStream = fs.createWriteStream('tag.csv', { flags: 'w' });
        csvStream.write('Autopista,RUT Cliente,Vence,Monto\n');

        for (const deuda of deudas) {
            const empresaInfo = await deuda.$('.empresa-info');
            const autopista = await empresaInfo.$eval('.nombre-s strong', el => el.textContent.trim());
            const rutCliente = await empresaInfo.$eval(':nth-child(2)', el => el.textContent.trim());
            const vence = await empresaInfo.$eval(':nth-child(4)', el => el.textContent.trim());

            // Intentar capturar el monto de manera segura
            let monto = 'No disponible';
            try {
                const montoElement = await deuda.$('.valor b');
                if (montoElement)
                    monto = await montoElement.textContent();
            } catch (error) {
                console.log(`No se pudo obtener el monto para ${autopista}`);
            }

            csvStream.write(`${autopista},${rutCliente},${vence},${monto}\n`);
        }

        csvStream.end();

        // Obtener el valor total a pagar
        const totalPagar = await page.$eval('.precio-desktop', el => el.textContent.trim());
        console.log('\nTotal a pagar:', totalPagar);

        // Mostrar desglose de deudas
        console.log('\nDesglose de deudas:');
        for (const deuda of deudas) {
            const empresaInfo = await deuda.$('.empresa-info');
            const autopista = await empresaInfo.$eval('.nombre-s strong', el => el.textContent.trim());
            let monto = 'No disponible';
            try {
                const montoElement = await deuda.$('.valor b');
                if (montoElement)
                    monto = await montoElement.textContent();
            } catch (error) {
                console.log(`No se pudo obtener el monto para ${autopista}`);
            }
            console.log(`${autopista}: ${monto}`);
        }

    } catch (error) {
        console.error('Ocurrió un error:', error);

    } finally {
        await browser.close();
    }
}

// Obtener el RUT como argumento de línea de comandos
if (process.argv.length !== 3) {
    console.log('Uso: node tag.js <RUT>');
    process.exit(1);
}

const rut = process.argv[2];
searchTagTotal(rut);
